src/runtime_src/core/common/api/xrt_kernel.cpp

```diff
@@ -2879,6 +2910,168 @@ class run::command_error_impl
+      throw xrt_core::error("runlist must be idle before adding run objects");
+
+
+    // Get the potentially throwing action out of the way first
+    m_runlist.reserve(m_runlist.size() + 1);
+    m_bos.reserve(m_bos.size() + 1);
+
+    auto run_impl = run.get_handle();
+    auto cmd = run_impl->get_cmd();
+    auto bo = cmd->get_exec_bo();
+
+    // Once a run object is added to a list it will be in a state that
+    // makes it impossible to add to another list or to same list
+    // twice.  This state is managed by the run object itself by
+    // recording this runlist with the run object, but it doesn't 
+    // proctect against caller manually controlling the run object,
+    // which is undefined behavior.  No exceptions after this point.
+    run_impl->set_runlist(this);  // throws or changes state of run
+
+    // No throw zone
+    m_runlist.push_back(std::move(run));  // move of shared_ptr is noexcept
+    m_bos.push_back(bo);                  // ptr noexcept
+  }
+
+  void
+  execute(const xrt::runlist& rl)
+  {
+    if (m_state != state::idle)
+      throw xrt_core::error("runlist must be idle before submitting for execution");
+
+    if (m_runlist.empty())
+      return;
+
+    // Prep each run object
+    for (auto& run : m_runlist)
+      run.get_handle()->prep_start();
+
+    // Close the command list.
+    m_state = state::closed;
+
+    // Submit
+    m_hwqueue.submit(rl);
+
+    // The command list is now submitted (running).  It cannot be reset
+    // until wait() has been called and state changed to idle or error.
+    m_state = state::running;
+  }
+
+  std::cv_status
+  wait_throw_on_error(const std::chrono::milliseconds& timeout)
+  {
+    if (m_state != state::running)
+      return std::cv_status::no_timeout;
+
+    if (m_state == state::error)
+      throw xrt_core::error("runlist is in error state and must be reset");
+
+    // Wait on last submitted command
+    auto last = m_runlist.back();
+
+    try {
+      auto status = last.wait2(timeout);
+      if (status == std::cv_status::no_timeout)
+        m_state = state::idle;
+
+      return status;
+    }
+    catch (const xrt::run::command_error& ex) {
+      // The runlist is in error state and must be reset before reuse.
+      m_state = state::error;
+
+      // Find first run with error
+      for (auto& run : m_runlist)
+        if (auto state = run.get_handle()->state(); state != ERT_CMD_STATE_COMPLETED)
+          throw xrt::runlist::command_error(run, state, ex.what());
+
+      throw xrt_core::error("internal error: no run with error found");
+    }
+  }
+
+  void
+  reset()
+  {
+    if (m_state == state::running)
+      throw xrt_core::error("runlist is submitted for execution and cannot be reset");
+
+    clear_runs();
+
+    m_runlist.clear();
+    m_bos.clear();
+    m_state = state::idle;
+  }
+};
+
+
+class runlist::command_error_impl
+{
+public:
+  xrt::run m_run;
+  ert_cmd_state m_state;
+  std::string m_message;
```

Comment:
Should these be `const` members?

Reply:
Why?  This are data members are part of a pimpl, opaque to exposed API.  Access is limited to API implementation within this source file, which does return by-value copies.